cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerHttpHandlerInternal.java

```diff
@@ -60,26 +63,65 @@
 public class TaskWorkerHttpHandlerInternal extends AbstractHttpHandler {
+  /**
+   * Fraction of duration which will be used for calculating a range.
+   * I.e., range = [Duration - DURATION_FRACTION * Duration, Duration + DURATION_FRACTION * Duration]
+   * A number (i.e., duration) will then get randomly selected within this range,
+   * and worker pod will get restarted after this randomly selected duration.
+   * Reason: by randomizing the duration, it is guaranteed that pods do not get restarted at the same time.
+   */
+  private static final double DURATION_FRACTION = 0.1;
   private static final Logger LOG = LoggerFactory.getLogger(TaskWorkerHttpHandlerInternal.class);
   private static final Gson GSON = new GsonBuilder().registerTypeAdapter(BasicThrowable.class,
                                                                          new BasicThrowableCodec()).create();
   private final RunnableTaskLauncher runnableTaskLauncher;
   private final BiConsumer<Boolean, String> stopper;
+
+  /**
+   * Holds the total number of requests that are being handled by this handler.
+   * Has the following invariant: inflightRequests should always be less than or equal to 1.
+   */
   private final AtomicInteger inflightRequests = new AtomicInteger(0);
+
+  /**
+   * Holds the total number of requests that have been executed by this handler that should count toward max allowed.
+   * Example, if a request is not valid, or does not involve user code, it should not be counted.
+   */
+  private final AtomicInteger totalRequestsWithUserCodeExecution = new AtomicInteger(0);
+
   private final String metadataServiceEndpoint;
 
   public TaskWorkerHttpHandlerInternal(CConfiguration cConf, Consumer<String> stopper) {
     this.runnableTaskLauncher = new RunnableTaskLauncher(cConf);
     this.metadataServiceEndpoint = cConf.get(Constants.TaskWorker.METADATA_SERVICE_END_POINT);
     this.stopper = (terminate, className) -> {
-      if (cConf.getBoolean(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION) && terminate) {
-        if (className != null) {
-          stopper.accept(className);
-        }
+      if (!terminate || className == null ||
+        cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION, 0) <= 0) {
+        // No need to restart.
+        totalRequestsWithUserCodeExecution.decrementAndGet();
+        inflightRequests.set(0);
+        return;
+      }
+
+      if (totalRequestsWithUserCodeExecution.get() >=
+        cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_EXECUTION)) {
+        stopper.accept(className);
       } else {
         inflightRequests.set(0);
       }
     };
+
+    int duration = cConf.getInt(Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION, 0);
+    int lowerBound = (int) (duration - duration * DURATION_FRACTION);
+    int upperBound = (int) (duration + duration * DURATION_FRACTION);
+    if (lowerBound > 0) {
+      int waitTime = (new Random()).nextInt(upperBound - lowerBound) + lowerBound;
```

Comment:
What is this for ?

Reply:
I think you should execute this time based restart logic first before resetting the inflight so that you don’t need to worry about shutting down the http server while there is inflight request

Reply:
We do not restart the pod after waiting precisely for Constants.TaskWorker.CONTAINER_KILL_AFTER_DURATION seconds. Reason: many pods will end up restarting at the same time. 
Instead, based on CONTAINER_KILL_AFTER_DURATION, we introduce a range and randomly select the wait time from this range. This ensures that not all pods get restarted at the same time.

Reply:
The time base restart logic is as follows: if there is no ongoing request, we increment inflightRequest counter so no new request gets accepted and restart the pod right away.
Otherwise, we wait until the ongoing request finishes. Once finished, we restart the pod irrespective of whether the last executed request needed a restart or not.